File: 2466-count-ways-to-build-good-strings/2466-count-ways-to-build-good-strings.cpp
class Solution {
public:
    int countGoodStrings(int low, int high, int zero, int one) {
        // Use dp[i] to record to number of good strings of length i.
        const int mod = 1'000'000'007;
        vector<long> dp(high + 1);
        dp[0] = 1;

        // Iterate over each length 'end'.
        for (int end = 1; end <= high; ++end){
            // Check if the current string can be made by append zero '0's or one '1's.
            if (end >= zero)
                dp[end] += dp[end - zero] % mod;
            if (end >= one)
                dp[end] += dp[end - one] % mod;

            dp[end] %= mod;
        }

        // Add up to the number of strings with each valid length [low ~ high].
        return accumulate(dp.begin() + low, dp.begin() + high + 1, 0L) % mod;
    }
};

File: 2466-count-ways-to-build-good-strings/2466-count-ways-to-build-good-strings.py
class Solution:
    def countGoodStrings(self, low: int, high: int, zero: int, one: int) -> int:
        def dfs(end: int) -> int:
            # Base case
            if dp[end] != -1:
                return dp[end]
            ans = 0
            if end >= zero:
                ans += dfs(end - zero)
            
            if end >= one:
                ans += dfs(end - one)
            dp[end] = ans % mod
            return dp[end]
        
        # Use dp[i] to record to number of good strings of length i.
        dp, mod = [-1] * (high + 1), 1_000_000_007
        dp[0] = 1

        # Add up the number of strings with each valid length [low ~ high]
        return sum(dfs(end) for end in range(low, high + 1)) % mod